#9410 - ConvertInputFormat incorrectly handles file name with period in name

jmarrec opened the pull request:
Pull request overview
---------------------
 - Fixes #9410


### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
Add some ctests for ConvertInputFormat. It uses Ctest "Fixtures", not sure how much I dislike it

jmarrec committed:
Fix #9410 - Correctly handle extension

jmarrec committed:
#9410 - Apply the same fix to E+ itself, and remove unused global variables

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 9410_ConvertInputFormat

jmarrec committed:
verify_cmake_dirs.py complains about CMAKE_SOURCE_DIR so use PROJECT_SOURCE_DIR/../

jmarrec commented on src/EnergyPlus/DataStringGlobals.hh:
```diff
@@ -113,19 +113,15 @@ struct DataStringGlobalsData : BaseGlobalStruct
+    fs::path inputFilePath; // Full name of file to open, including path `/path/to/myfile.idf`
     fs::path inputIddFilePath;
     fs::path inputEpJSONSchemaFilePath;
-    fs::path idfDirPath;
```
I also cleaned up a bunch of unused global variables

Myoldmopar replied:
:heart:

jmarrec commented on src/EnergyPlus/DataStringGlobals.hh:
```diff
@@ -113,19 +113,15 @@ struct DataStringGlobalsData : BaseGlobalStruct
     fs::path eplusADSFilePath;
-    fs::path weatherFilePathNameOnly;
-    fs::path inputFilePath;
+    fs::path inputFilePath; // Full name of file to open, including path `/path/to/myfile.idf`
```
clarify this

jmarrec commented on src/EnergyPlus/DataStringGlobals.hh:
```diff
@@ -113,19 +113,15 @@ struct DataStringGlobalsData : BaseGlobalStruct
-    fs::path inputDirPath;
-    fs::path inputFilePathNameOnly;
+    fs::path inputDirPath;          // `/path/to/`
+    fs::path inputFilePathNameOnly; // `myfile.idf`
```
and this

jmarrec commented on src/ConvertInputFormat/main.cpp:
```diff
@@ -356,33 +354,33 @@ bool processInput(std::string const &inputFilePath,
     }
 
-    fs::path fileNameWithoutExtension = outputDirPath / inputFilePathNameOnly;
+    auto const outputFilePathWithOriExtension = outputDirPath / EnergyPlus::FileSystem::getFileName(inputFilePath);
```
The bottom line of the changes is best understood by looking at this Compiler explorer: https://godbolt.org/z/Kvqqq1xnc

jmarrec commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -292,7 +292,7 @@ namespace CommandLineInterface {
         }
 
-        state.dataStrGlobals->inputFilePathNameOnly = FileSystem::removeFileExtension(FileSystem::getFileName(state.dataStrGlobals->inputFilePath));
+        state.dataStrGlobals->inputFilePathNameOnly = FileSystem::getFileName(state.dataStrGlobals->inputFilePath);
```
Same fix for E+'s CLI itself

Myoldmopar replied:
OK great!

jmarrec commented on src/ConvertInputFormat/CMakeLists.txt:
```diff
@@ -26,3 +26,67 @@ set_target_properties(ConvertInputFormat PROPERTIES VERSION ${ENERGYPLUS_VERSION
+  add_test(NAME ConvertInputFormat.RegularFile_AbsolutePath.TestOK
+    COMMAND ${CMAKE_COMMAND} -E cat "${TEST_DIR}/RegularFile_AbsolutePath/1ZoneUncontrolled.epJSON" # cat will fail if it doesn't exist
+  )
+  set_tests_properties(ConvertInputFormat.RegularFile_AbsolutePath.TestOK PROPERTIES FIXTURES_CLEANUP RegularFile_AbsolutePathFixture)
+  set_tests_properties(ConvertInputFormat.RegularFile_AbsolutePath PROPERTIES FIXTURES_REQUIRED "ConvertInputFormat.SetupFixture; RegularFile_AbsolutePathFixture")
+
+
+  add_test(NAME ConvertInputFormat.RegularFile_RelativePath
+    COMMAND ConvertInputFormat --output RegularFile_RelativePath "1ZoneUncontrolled.idf"
+    WORKING_DIRECTORY ${TEST_DIR}
+  )
+  add_test(NAME ConvertInputFormat.RegularFile_RelativePath.TestOK
+    COMMAND ${CMAKE_COMMAND} -E cat "${TEST_DIR}/RegularFile_RelativePath/1ZoneUncontrolled.epJSON" # cat will fail if it doesn't exist
+  )
+  set_tests_properties(ConvertInputFormat.RegularFile_RelativePath.TestOK PROPERTIES FIXTURES_CLEANUP RegularFile_RelativePathFixture)
+  set_tests_properties(ConvertInputFormat.RegularFile_RelativePath PROPERTIES FIXTURES_REQUIRED "ConvertInputFormat.SetupFixture; RegularFile_RelativePathFixture")
+
+
+
+  add_test(NAME ConvertInputFormat.ExtraDotFile_AbsolutePath
+    COMMAND ConvertInputFormat --output "${TEST_DIR}/ExtraDotFile_AbsolutePath" "${TEST_DIR}/1.ZoneUncontrolled.idf"
+  )
+  add_test(NAME ConvertInputFormat.ExtraDotFile_AbsolutePath.TestOK
+    COMMAND ${CMAKE_COMMAND} -E cat "${TEST_DIR}/ExtraDotFile_AbsolutePath/1.ZoneUncontrolled.epJSON" # cat will fail if it doesn't exist
+  )
+set_tests_properties(ConvertInputFormat.ExtraDotFile_AbsolutePath.TestOK PROPERTIES FIXTURES_CLEANUP ExtraDotFile_AbsolutePathFixture)
+  set_tests_properties(ConvertInputFormat.ExtraDotFile_AbsolutePath PROPERTIES FIXTURES_REQUIRED "ConvertInputFormat.SetupFixture; ExtraDotFile_AbsolutePathFixture")
+
+
+  add_test(NAME ConvertInputFormat.ExtraDotFile_RelativePath
+    COMMAND ConvertInputFormat --output ExtraDotFile_RelativePath "1.ZoneUncontrolled.idf"
+    WORKING_DIRECTORY ${TEST_DIR}
+  )
+  add_test(NAME ConvertInputFormat.ExtraDotFile_RelativePath.TestOK
+    COMMAND ${CMAKE_COMMAND} -E cat "${TEST_DIR}/ExtraDotFile_RelativePath/1.ZoneUncontrolled.epJSON" # cat will fail if it doesn't exist
+  )
+  set_tests_properties(ConvertInputFormat.ExtraDotFile_RelativePath.TestOK PROPERTIES FIXTURES_CLEANUP ExtraDotFile_RelativePathFixture)
+  set_tests_properties(ConvertInputFormat.ExtraDotFile_RelativePath PROPERTIES FIXTURES_REQUIRED "ConvertInputFormat.SetupFixture; ExtraDotFile_RelativePathFixture")
+
+endif()
```
Added some ctest for ConvertInputFormat... This makes use of CTest **Fixtures**. This isn't terribly fun to read but I'm just basically doing

* Setup fixtures: remove any existing dir, remake dirs. Then copy 1ZoneUncontrolled to the test dir
* Run ConvertInput format on that file
* Cleanup fixture: asserts that the epJSON created has the correct name by using `cmake -E cat` on it (that's a cross platform workaround...).

jmarrec committed:
Wrong relative path...

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 9410_ConvertInputFormat

jmarrec commented:
I am confused by why mac and windows CI have some problems with the test. I just tested locally on a Windows 10 (cmake 3.22.1) and macOS M1 (cmake 3.23.1) in addition to Ubuntu 20.04 (cmake 3.23.1) and it all works.

```shell
cmake .
cmake --build . --target ConvertInputFormat --config Release
ctest -R ConvertInputFormat
```

Myoldmopar commented:
Is it a problem with the CMake Fixture stuff?  Would it be worthwhile to temporarily just repeat the setup/cleanup logic instead of using a fixture to see if CI is happier?  Then at least we'd be able to eliminate that as an issue if that doesn't fix it.  Otherwise I'm not sure why it's not working right now.

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 9410_ConvertInputFormat

jmarrec committed:
Move to a Python-based ctest instead of Ctest Fixtures: CI doesn't seem to be able to run CTest fixtures correctly (I tested locally with cmake 3.22+ on mac, linux and ubuntu...)

My guess is that the cmake version on CI is too old

jmarrec commented:
@Myoldmopar I removed the use of CMake fixtures. I'm guessing this is a problem with the cmake version used on the runners which may be too old. I moved to a python script. I am now very confused about  what the Windows runner is now doing wrong with that script... It seems it cannot understand f-strings, which are a python 3.6 features (which is the minimum version we enforce...)

Any ideas please?

It's complaining:

```
 File "", line 1
(convertinputformat_exe=)
^
SyntaxError: invalid syntax
```

I guess (?) it's coming from here: https://github.com/NREL/EnergyPlus/blob/90805271736d334ceeb9762276408010a94de0f6/src/ConvertInputFormat/test_convert_input_format.py#L117

Myoldmopar commented:
Yes, the f strings were added in 3.6, but your equals operator voodoo inside the f string evaluation braces wasn't added [until 3.8](https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging).

jmarrec committed:
Remove {fstring=} as windows is using python 3.6 and this wasn't added until 3.8

jmarrec commented:
@Myoldmopar I didn't realize this wasn't until 3.8, thanks. I made the change. I guess windows CI is using 3.6 then.

(It may be getting a bit weirdish to require python 3.6, but forcing 3.8 when using LINK_PYTHON on github actions)

Myoldmopar commented:
CI looks all happy now, I'll take a final look at the code shortly.

Myoldmopar commented on src/ConvertInputFormat/CMakeLists.txt:
```diff
@@ -26,3 +26,47 @@ set_target_properties(ConvertInputFormat PROPERTIES VERSION ${ENERGYPLUS_VERSION
+             --input-file "1.ZoneUncontrolled.idf"
+    WORKING_DIRECTORY ${TEST_DIR}
+  )
+endif()
```
Great!

Myoldmopar reviewed (approved):
If you take out the test code, this appears to be a reduction of 4 lines of code, with a great fix on the file name.  The tests make sense and cover the defect.  I'll build and test locally but I think this is ready.

Myoldmopar commented:
Built locally, and encountered the original issue on develop, with a properly converted file name using ConvertInputFormat on this branch.  I think this is ready to go in, merging this one.  Thanks @jmarrec !
